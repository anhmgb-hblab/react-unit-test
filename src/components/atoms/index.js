import AButton from './AButton';
import AInput from './AInput';
import ACheckbox from './ACheckbox';
import AImage from './AImage';
import ATypography from './ATypography';
import ACard from './ACard';

export {
  AButton,
  AInput,
  ACheckbox,
  AImage,
  ATypography,
  ACard,
};
