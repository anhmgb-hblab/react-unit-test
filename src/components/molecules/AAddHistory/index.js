import React from 'react';
import { Divider } from 'antd';
import PropTypes from 'prop-types';
import {
  AInput,
  ACard,
  AButton,
  ASelect,
  ARadio,
  ADatepicker,
} from '../../atoms';
import { HISTORY_TYPE, HISTORY_OWNER } from '../../../constants/format';
import StyledAAddHistory from './styled';

const AAddHistory = ({
  newHistory,
  onAddHistory,
  onCancelHistory,
  onChangeNewHistory,
}) => {
  if (!newHistory) return null;
  return (
    <StyledAAddHistory>
      <ACard>
        <div className="add-history">
          <ARadio
            className="radio"
            label="History type"
            options={HISTORY_TYPE}
            value={newHistory.type}
            onChange={(e) => onChangeNewHistory('type', e.target.value)}
          />
          <ASelect
            className="select"
            label="History owner"
            options={HISTORY_OWNER}
            value={newHistory.owner}
            onChange={(e) => onChangeNewHistory('owner', e.value)}
          />
          <ADatepicker
            className="date"
            label="Date"
            value={newHistory.date}
            onChange={(value) => onChangeNewHistory('date', value)}
          />
          <AInput
            className="amount"
            label="Amount"
            value={newHistory.amount}
            onChange={(e) => onChangeNewHistory('amount', e.target.value)}
          />
          <AInput
            className="reason"
            label="Reason"
            value={newHistory.reason}
            onChange={(e) => onChangeNewHistory('reason', e.target.value)}
          />
        </div>
        <Divider />
        <div className="add-history__actions">
          <AButton type="default" onClick={onCancelHistory}>
            Cancel
          </AButton>
          <AButton onClick={onAddHistory}>
            Save
          </AButton>
        </div>
      </ACard>
    </StyledAAddHistory>
  );
};

AAddHistory.defaultProps = {
  newHistory: null,
};

AAddHistory.propTypes = {
  newHistory: PropTypes.objectOf(PropTypes.any),
  onAddHistory: PropTypes.func.isRequired,
  onCancelHistory: PropTypes.func.isRequired,
  onChangeNewHistory: PropTypes.func.isRequired,
};

export default AAddHistory;
