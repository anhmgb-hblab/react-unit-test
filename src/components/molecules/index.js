import ATodo from './ATodo';

const foo = {};

export { ATodo, foo };
