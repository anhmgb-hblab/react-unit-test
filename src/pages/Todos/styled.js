import styled from 'styled-components';

export default styled.div`
  margin: 200px auto;
  width: 400px;

  .todo {
    display: flex;
    flex-direction: column;
    margin: 0 auto;
  }

  .a-input {
    margin-bottom: 8px;
  }

  .a-todo {
    margin-bottom: 8px;
  }
`;
