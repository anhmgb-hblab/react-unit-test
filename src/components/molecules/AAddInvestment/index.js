import React from 'react';
import { Divider } from 'antd';
import PropTypes from 'prop-types';
import { AInput, ACard, AButton } from '../../atoms';
import StyledAAddInvestment from './styled';

const AAddInvestment = ({
  newInvestment,
  onAddInvestment,
  onCancelInvestment,
  onChangeNewInvestment,
}) => {
  if (!newInvestment) return null;
  return (
    <StyledAAddInvestment>
      <ACard>
        <div className="add-investment">
          <AInput
            label="Investment plan name"
            value={newInvestment.planName}
            onChange={(e) => onChangeNewInvestment(newInvestment.id, 'planName', e.target.value)}
          />
          <AInput
            label="Investment Target"
            value={newInvestment.investmentTarget}
            onChange={(e) => onChangeNewInvestment(newInvestment.id, 'investmentTarget', e.target.value)}
          />
          <AInput
            label="Due plan"
            value={newInvestment.yearTarget}
            onChange={(e) => onChangeNewInvestment(newInvestment.id, 'yearTarget', e.target.value)}
          />
          <AInput
            label="Interest rate"
            value={newInvestment.interestRate}
            onChange={(e) => onChangeNewInvestment(newInvestment.id, 'interestRate', e.target.value)}
          />
          <AInput
            label="Init self assets"
            value={newInvestment.initSelfAssets}
            onChange={(e) => onChangeNewInvestment(newInvestment.id, 'initSelfAssets', e.target.value)}
          />
          <AInput
            label="Init parent assets"
            value={newInvestment.initParentAssets}
            onChange={(e) => onChangeNewInvestment(newInvestment.id, 'initParentAssets', e.target.value)}
          />
          <AInput
            label="Init income"
            value={newInvestment.initIncome}
            onChange={(e) => onChangeNewInvestment(newInvestment.id, 'initIncome', e.target.value)}
          />
          <AInput
            label="Accumulate income per year"
            value={newInvestment.accumulateIncomePerYear}
            onChange={(e) => onChangeNewInvestment(newInvestment.id, 'accumulateIncomePerYear', e.target.value)}
          />
        </div>
        <Divider />
        <div className="add-investment__actions">
          <AButton type="default" onClick={onCancelInvestment}>
            Cancel
          </AButton>
          <AButton onClick={onAddInvestment}>
            Save
          </AButton>
        </div>
      </ACard>
    </StyledAAddInvestment>
  );
};

AAddInvestment.propTypes = {
  newInvestment: PropTypes.objectOf(PropTypes.string).isRequired,
  onAddInvestment: PropTypes.func.isRequired,
  onCancelInvestment: PropTypes.func.isRequired,
  onChangeNewInvestment: PropTypes.func.isRequired,
};

export default AAddInvestment;
