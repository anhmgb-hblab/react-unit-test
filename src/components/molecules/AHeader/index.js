import React from 'react';
import { useHistory } from 'react-router-dom';
import StyledHeader from './styled';
import { AImage } from '../../atoms';
import Logo from '../../../assets/images/investment-logo.jpeg';

const AHeader = () => {
  const history = useHistory();

  const handleRedirectToDashboard = () => {
    history.push('/');
    window.location.reload();
  };

  return (
    <StyledHeader>
      <div className="header__logo">
        <AImage
          src={Logo}
          alt="logo"
          onClick={handleRedirectToDashboard}
        />
      </div>
    </StyledHeader>
  );
};

export default AHeader;
