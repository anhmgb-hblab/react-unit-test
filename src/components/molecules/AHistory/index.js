import React from 'react';
import { Collapse } from 'antd';
import { CaretUpOutlined, CaretDownOutlined } from '@ant-design/icons';
import PropTypes from 'prop-types';
import { ATypography, ATag } from '../../atoms';
import { HISTORY_TYPE, HISTORY_OWNER } from '../../../constants/format';
import StyledAHistory, { StyledIcon } from './styled';

const { Panel } = Collapse;

const AHistory = ({ histories }) => {
  const renderHistoryHeader = (history) => {
    const {
      type,
      amount,
      owner,
      date,
    } = history;
    const isTypeUp = type === HISTORY_TYPE[0].value;
    const isOwnerSelf = owner === HISTORY_OWNER[0].value;

    const renderIcon = () => {
      if (isTypeUp) return <CaretUpOutlined />;
      return <CaretDownOutlined />;
    };

    return (
      <div className="history-item">
        <div className="history-item__left">
          <div className="history-item__left__amount">
            <StyledIcon isTypeUp={isTypeUp}>
              {renderIcon()}
            </StyledIcon>
            <ATypography type={isTypeUp ? 'up' : 'down'}>
              {parseInt(amount, 10).toLocaleString()}
            </ATypography>
          </div>
          <ATag type="date" value={date} />
        </div>
        <div className="history-item__right">
          <ATag
            type={isOwnerSelf ? 'self' : 'parent'}
            value={HISTORY_OWNER[owner - 1].label}
          />
        </div>
      </div>
    );
  };

  if (!histories) return null;
  return (
    <StyledAHistory>
      {
        histories.map((history) => (
          <Collapse collapsible="header" key={history.id}>
            <Panel
              showArrow={false}
              header={renderHistoryHeader(history)}
              key={history.id}
            >
              {history.reason}
            </Panel>
          </Collapse>
        ))
      }
    </StyledAHistory>
  );
};

AHistory.defaultProps = {
  histories: null,
};

AHistory.propTypes = {
  histories: PropTypes.arrayOf(PropTypes.any),
};

export default AHistory;
