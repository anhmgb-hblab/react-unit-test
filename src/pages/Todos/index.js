import React, { useState } from 'react';
import { v4 as uuidv4 } from 'uuid';
import { AInput } from '../../components/atoms';
import { ATodo } from '../../components/molecules';
import StyledTodos from './styled';

const Todos = () => {
  const [newTodoContent, setNewTodoContent] = useState('');
  const [todos, setTodos] = useState([]);

  const handleAddNewTodo = () => {
    const newTodo = {
      id: uuidv4(),
      content: newTodoContent,
      checked: false,
    };
    setTodos([newTodo, ...todos]);
    setNewTodoContent('');
  };

  const handleChangeTodoStatus = (id) => {
    const newTodos = todos.map((todo) => {
      if (todo.id !== id) return todo;
      return {
        ...todo,
        checked: !todo.checked,
      };
    });
    setTodos(newTodos);
  };

  const handleDeleteTodo = (id) => {
    const newTodos = todos.filter((items) => items.id !== id);
    setTodos(newTodos);
  };

  const renderTodos = () => {
    if (!todos.length) return null;
    return (
      todos.map((todo) => (
        <ATodo
          className="todos__item"
          key={todo.id}
          content={todo.content}
          checked={todo.checked}
          onChangeStatus={() => handleChangeTodoStatus(todo.id)}
          onDeleteTodo={() => handleDeleteTodo(todo.id)}
        />
      ))
    );
  };

  return (
    <StyledTodos>
      <div className="todos">
        <AInput
          value={newTodoContent}
          onChange={(e) => setNewTodoContent(e.target.value)}
          placeholder="Input your new todo"
          onPressEnter={handleAddNewTodo}
        />
        {renderTodos()}
      </div>
    </StyledTodos>
  );
};

export default Todos;
