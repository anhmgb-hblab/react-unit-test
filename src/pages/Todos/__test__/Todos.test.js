import React from 'react';
import toJson from 'enzyme-to-json';
import { mount } from 'enzyme';
import Todos from '../index';
import { ATodo } from '../../../components/molecules';

describe('Test Todos running correctly', () => {
  it('test onChange input', () => {
    const wrapper = mount(<Todos />);
    const input = wrapper.find('input').first();
    input.simulate('focus');
    input.simulate('change', { target: { value: 'Todo' } });
    expect(input.render().attr('value')).toBe('Todo');
  });

  it('test add new todo', async () => {
    const setState = jest.fn();
    const useStateSpy = jest.spyOn(React, 'useState');
    useStateSpy.mockImplementation(() => [[], setState]);

    const wrapper = mount(<Todos />);
    const input = wrapper.find('input').first();
    input.simulate('focus');
    input.simulate('change', { target: { value: 'Todo' } });
    input.simulate('keyDown', { keyCode: 13 });
    wrapper.update();

    const items = wrapper.find(ATodo);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(items).toHaveLength(1);
  });
});
