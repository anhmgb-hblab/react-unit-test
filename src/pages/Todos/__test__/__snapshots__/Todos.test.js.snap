// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Todos running correctly test add new todo 1`] = `
<Todos>
  <styled.div>
    <div
      className="sc-dlfnuX hCfXGJ"
    >
      <div
        className="todos"
      >
        <AInput
          allowClear={false}
          autoSize={false}
          disabled={false}
          label=""
          onChange={[Function]}
          onFocus={[Function]}
          onPressEnter={[Function]}
          placeholder="Input your new todo"
          readonly={false}
          size="middle"
          type="text"
          value=""
        >
          <styled.div
            className="a-input"
          >
            <div
              className="sc-bdfBQB fLvfYJ a-input"
            >
              <Input
                allowClear={false}
                disabled={false}
                onChange={[Function]}
                onFocus={[Function]}
                onPressEnter={[Function]}
                placeholder="Input your new todo"
                readOnly={false}
                size="middle"
                type="text"
                value=""
              >
                <ClearableLabeledInput
                  allowClear={false}
                  bordered={true}
                  disabled={false}
                  element={
                    <input
                      className="ant-input"
                      disabled={false}
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      placeholder="Input your new todo"
                      readOnly={false}
                      type="text"
                      value=""
                    />
                  }
                  focused={true}
                  handleReset={[Function]}
                  inputType="input"
                  onChange={[Function]}
                  onFocus={[Function]}
                  onPressEnter={[Function]}
                  placeholder="Input your new todo"
                  prefixCls="ant-input"
                  readOnly={false}
                  size="middle"
                  triggerFocus={[Function]}
                  type="text"
                  value=""
                >
                  <input
                    className="ant-input"
                    disabled={false}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    placeholder="Input your new todo"
                    readOnly={false}
                    type="text"
                    value=""
                  />
                </ClearableLabeledInput>
              </Input>
            </div>
          </styled.div>
        </AInput>
        <ATodo
          checked={false}
          className="todos__item"
          content="Todo"
          key="f24e3d02-fdfc-4345-9c9c-7f4e59621e84"
          onChangeStatus={[Function]}
        >
          <styled.div
            className="a-todo"
          >
            <div
              className="sc-gsTEea kUzTQc a-todo"
            >
              <div
                className="content"
              >
                Todo
              </div>
              <ACheckbox
                checked={false}
                disabled={false}
                label={null}
                onChange={[Function]}
              >
                <Checkbox
                  checked={false}
                  disabled={false}
                  onChange={[Function]}
                >
                  <label
                    className="ant-checkbox-wrapper"
                  >
                    <Checkbox
                      checked={false}
                      className=""
                      defaultChecked={false}
                      disabled={false}
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      onKeyPress={[Function]}
                      onKeyUp={[Function]}
                      prefixCls="ant-checkbox"
                      style={Object {}}
                      type="checkbox"
                    >
                      <span
                        className="ant-checkbox"
                        style={Object {}}
                      >
                        <input
                          checked={false}
                          className="ant-checkbox-input"
                          disabled={false}
                          onBlur={[Function]}
                          onChange={[Function]}
                          onFocus={[Function]}
                          onKeyDown={[Function]}
                          onKeyPress={[Function]}
                          onKeyUp={[Function]}
                          type="checkbox"
                        />
                        <span
                          className="ant-checkbox-inner"
                        />
                      </span>
                    </Checkbox>
                    <span />
                  </label>
                </Checkbox>
              </ACheckbox>
            </div>
          </styled.div>
        </ATodo>
      </div>
    </div>
  </styled.div>
</Todos>
`;
